uefi-pi/src/pi/guid.rs

```diff
@@ -0,0 +1,95 @@
+use core::{convert::TryInto, mem::size_of, ptr::slice_from_raw_parts};
+
+const GUID_STRING_LEN: usize = 36;
+
+// A GUID is a 128-bit integer (16 bytes) that can be
+// used as a unique identifier.
+#[repr(C)]
+#[derive(Debug, Default)]
+pub struct Guid {
+    f0: u32,
+    f1: u16,
+    f2: u16,
+    f3: [u8; 8],
+}
+
+#[derive(Debug)]
+pub enum GuidParseError {
+    InvalidInput,
+}
+
+impl Guid {
+    // Create a GUID instance from several fields
+    pub const fn from_fields(f0: u32, f1: u16, f2: u16, f3: [u8; 8]) -> Guid {
+        Self { f0, f1, f2, f3 }
+    }
+
+    pub fn as_bytes(&self) -> &[u8; 16] {
+        // Safe since the size of Guid is 16
+        unsafe {
+            (&*slice_from_raw_parts(self as *const Self as *const u8, size_of::<Self>()))
+                .try_into()
+                .unwrap()
+        }
+    }
+
+    // Create a GUID instance from a string slice
+    // Input should follow format strictly: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
+    // For example: "F9168C5E-CEB2-4faa-B6BF-329BF39FA1E4"
+    pub fn parse_str(s: &str) -> Result<Self, GuidParseError> {
+        // Make sure the input lenght is legal
+        if s.len() != GUID_STRING_LEN {
+            return Err(GuidParseError::InvalidInput);
+        }
+        let parse_err = |_| GuidParseError::InvalidInput;
+
+        // Parse the string into fields
+        let f0 = u32::from_str_radix(&s[0..8], 16).map_err(parse_err)?;
+        let f1 = u16::from_str_radix(&s[9..13], 16).map_err(parse_err)?;
+        let f2 = u16::from_str_radix(&s[14..18], 16).map_err(parse_err)?;
+        let mut f3 = u64::from_str_radix(&s[19..23], 16).map_err(parse_err)? << 48;
+        f3 |= u64::from_str_radix(&s[24..36], 16).map_err(parse_err)?;
+
+        // f3 is decoded from string so use big endian to encode into bytes
+        Ok(Self {
+            f0,
+            f1,
+            f2,
+            f3: u64::to_be_bytes(f3),
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_guid() {
+        // F9168C5E-CEB2-4faa-B6BF-329BF39FA1E4
+        let guid_bytes = [
+            0x5E, 0x8C, 0x16, 0xF9, 0xB2, 0xCE, 0xaa, 0x4f, 0xB6, 0xBF, 0x32, 0x9B, 0xF3, 0x9F,
+            0xA1, 0xE4,
+        ];
+        let guid_field = Guid::from_fields(
+            0xF9168C5E,
+            0xCEB2,
+            0x4faa,
+            [0xB6, 0xBF, 0x32, 0x9B, 0xF3, 0x9F, 0xA1, 0xE4],
+        );
+
+        assert_eq!(&guid_bytes, guid_field.as_bytes());
+
+        let guid_str = Guid::parse_str("F9168C5E-CEB2-4faa-B6BF-329BF39FA1E4").unwrap();
```

Comment:
add a case for 
`F9168C5E_CEB2_4faa-B6BF-329BF39FA1E4`
and 
`F9168C5ECCEB2C4faaCB6BFC329BF39FA1E4`?